### ISSUE TITLE
"not enough bytes" when .elmi file write is interrupted by failure

### ISSUE THREAD

**Issue by evancz** [elm/compiler#2008]
It may be possible that `elm make` partially writes a `.elmi` file but gets interrupted when another module fails. This would leave the cached files in an unreadable state. It is **possible** that this can still happen with 0.19.1, but it has not been confirmed yet.

This was explained by @martenbr in https://github.com/elm/compiler/issues/1837#issuecomment-439845258:

* * *

As @mbark describes above the problem happens when invoking a simple elm make command twice. An interesting observation however is that the size of the .elmi file does vary between runs:

```
-rw-rw-r--. 1 martenb martenb 953846 Nov 19 09:52 elm-stuff/0.19.0/Widget-RecTable.elmi
-rw-rw-r--. 1 martenb martenb 1084847 Nov 19 09:51 elm-stuff/0.19.0/Widget-RecTable.elmi
-rw-rw-r--. 1 martenb martenb 691839 Nov 19 09:52 elm-stuff/0.19.0/Widget-RecTable.elmi
```

This indicates a concurrency issue so i did a strace and saw the following:
- The compiler seems to compile things in parallel (at the very least it is writing the .elmi in parallel)
- The compiler keeps writing to the broken elmi file even when it start printing the compilation error.
- The compiler exits when it has finished printing the compilation error, it keeps writing to the elmi file until this point but it never calls close on it.
- Printing the compilation error is a relatively slow operation

In other words it seems that to trigger this issue you need to have a very big file (that compiles fine) which the compiler is not able to finish writing during the time it is printing the compilation error. This would explain why only large projects are affected.

**Comment by evancz** [comment]
I walked through the code to see if this seems possible with 0.19.1.

The build process compiles each module on a separate light-weight thread with `forkIO`. These threads may write two files:

- [`.elmi`](https://github.com/elm/compiler/blob/0.19.1/builder/src/Build.hs#L730)
- [`.elmo`](https://github.com/elm/compiler/blob/0.19.1/builder/src/Build.hs#L719)

If these threads were killed partway through, it would lead to a partial file write as described by @martenbr. The question is, can these threads be killed by a normal compile error?

The `checkModule` function is called with [this convention](https://github.com/elm/compiler/blob/0.19.1/builder/src/Build.hs#L214):

```haskell                  
  resultsMVars <- forkWithKey (checkModule env foreigns rmvar) statuses
  putMVar rmvar resultsMVars
  results <- traverse readMVar resultsMVars
```

This means that:

1. A thread is created for each module.
2. The threads do not start computation until all threads are spawned.
3. We block until all threads have completed.

That means **a normal compilation error (like a type error) will not lead to any corrupted files.**

I am not sure what happens if you encounter a compiler bug that causes a crash though. It is conceivable to me that this could lead to corrupt files. It should not happen in normal scenarios though as it seemed to be in #1837.

**Comment by evancz** [comment]
If you are encountering something that looks like this, **please open a new issue.**

Tell us your editor, your OS, any "file watcher" you are using, and any other information you can think of about your development environment!
